src/Storages/StorageMerge.cpp

```diff
@@ -658,6 +700,19 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
         StorageView * view = dynamic_cast<StorageView *>(storage.get());
         if (!view || allow_experimental_analyzer)
         {
+            row_policy_data.init(
+                modified_context->getRowPolicyFilter(
+                    database_name,
+                    table_name,
+                    RowPolicyFilterType::SELECT_FILTER),
+                storage,
+                modified_context);
```

Comment:
If you move the code from `RowPolicyData::init()` to the constructor `RowPolicyData::RowPolicyData()` and then change the type of the `row_policy_data` variable to `std::optional<RowPolicyData>` then the code will look cleaner and you'll able to get rid of the function `RowPolicyData::hasRowPolicy()`

Reply:
I doubt that suggested approach leads to cleaner code, because the decision if hasRowPolicy or not will be moved to outer scope.
Nevertheless, I will do it.